core-java-modules/core-java-concurrency-collections-2/src/main/java/com/baeldung/concurrent/lock/SingleLock.java

```diff
@@ -0,0 +1,46 @@
+package com.baeldung.concurrent.lock;
+
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantLock;
+
+import com.google.common.base.Supplier;
+
+public class SingleLock extends ConcurrentAccessExperiment {
+    ReentrantLock lock;
+
+    public SingleLock() {
+        lock = new ReentrantLock();
+    }
+
+    protected Supplier<?> putSupplier(Map<String,String> map, int key) {
+        return (()-> {
+            try {
+                lock.lock();
+                map.put("key" + key, "value" + key);
+            } catch (Exception e) {
+                this.putSupplier(map, key);
```

Comment:
Wouldn't this be an infinite loop? Seems risky to just infinitely try again. I'd recommend doing:

```java
lock.lock();
try {
    return map.put("key" + key, "value" + key);
} finally {
    lock.unlock();
}
```
This is what the JavaDoc recommends.

Also, if you return the result of `map.put`, then you don't need the extra `return null;` at the bottom.

Reply:
Well, what if the lock is not acquired? the thread won't execute and we lose a put/get

Reply:
`lock` will block until the lock is acquired. From the JavaDoc:

> If the lock is held by another thread then the current thread becomes disabled for thread scheduling purposes and lies dormant until the lock has been acquired, at which time the lock hold count is set to one.

Reply:
The thread will block until the lock is acquired. From the JavaDoc:

> If the lock is held by another thread then the current thread becomes disabled for thread scheduling purposes and lies dormant until the lock has been acquired, at which time the lock hold count is set to one.

I guess the question I have is, what is so different about this use case that we'd deviate from the JavaDoc?

Reply:
Ok